agent/consul/leader.go

```diff
@@ -213,17 +230,21 @@ WAIT:
 	for {
 		select {
 		case <-stopCh:
-			return
+			return nil
 		case <-s.shutdownCh:
-			return
+			return nil
 		case <-interval:
 			goto RECONCILE
 		case member := <-reconcileCh:
 			s.reconcileMember(member)
 		case index := <-s.tombstoneGC.ExpireCh():
 			go s.reapTombstones(index)
 		case errCh := <-s.reassertLeaderCh:
-			errCh <- reassert()
+			err := reassert()
+			errCh <- err
+			if err != nil {
+				return err
```

Comment:
This is the second usecase: we couldn't reassert and instead of retrying it after `interval`, an error is returned which leads to a raft leadership transfer.

Reply:
Hmmm, when does this happen? I don't recall the specifics and wonder if by failing and stepping down we might end up causing new leadership stability issues?

Reply:
https://github.com/hashicorp/consul/blob/09a1a32cfd57cfb83c08f0d7d89e40a92a06a138/agent/consul/server.go#L245-L246

This happens after snapshot restore. Since the agent revokes leadership and immediately tries to establish it again, there is the possibility that it fails. When it does we are in the same situation as above - raft thinks this agent is a leader, but consul disagrees.